### PR TITLE
[SPARK-35543][CORE][FOLLOWUP] Fix memory leak in BlockManagerMasterEndpoint removeRdd

### PR DESCRIPTION
**Pull request by attilapiros**
### What changes were proposed in this pull request?

Wrapping `JHashMap[BlockId, BlockStatus]` (used in `blockStatusByShuffleService`) into a new class `BlockStatusPerBlockId` which removes the reference to the map when all the persisted blocks are removed.

### Why are the changes needed?

With https://github.com/apache/spark/pull/32790 a bug is introduced when all the persisted blocks are removed we remove the HashMap which already shared by the block manger infos but when new block is persisted this map is needed to be used again for storing the data (and this HashMap must be the same which shared by the block manger infos created for registered block managers running on the same host where the external shuffle service is). 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Extending `BlockManagerInfoSuite` with test which removes all the persisted blocks then adds another one.

### CODE REVIEW

**Comment by attilapiros** [comment]
cc @mridulm

**Comment by attilapiros** [comment]
build error is unrelated:
```
[error] ## Exception when compiling 480 sources to /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/catalyst/target/scala-2.12/classes
[error] java.lang.StackOverflowError
[error] scala.reflect.api.Trees$Transformer.transform(Trees.scala:2563)
[error] scala.tools.nsc.transform.CleanUp$CleanUpTransformer.transform(CleanUp.scala:607)
[error] scala.tools.nsc.transform.CleanUp$CleanUpTransformer.transform(CleanUp.scala:37)
[error] scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2597)
[error] scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2595)
[error] scala.reflect.internal.Trees.itransform(Trees.scala:1430)
[error] scala.reflect.internal.Trees.itransform$(Trees.scala:1400)
```

Jenkins retest this please

**Comment by gengliangwang** [comment]
@attilapiros Thanks for coming up with a better fix.
cc @Ngone51 as well

**Review by mridulm — APPROVED** [review]
Thanks for fixing this @attilapiros !

**Review by dongjoon-hyun — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 1]
```diff
@@ -570,7 +565,7 @@ class BlockManagerMasterEndpoint(
       val externalShuffleServiceBlockStatus =
         if (externalShuffleServiceRddFetchEnabled) {
           val externalShuffleServiceBlocks = blockStatusByShuffleService
-            .getOrElseUpdate(externalShuffleServiceIdOnHost(id), new JHashMap[BlockId, BlockStatus])
+            .getOrElseUpdate(externalShuffleServiceIdOnHost(id), new BlockStatusPerBlockId)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:568]: Seems like we never clear the key after this change. Could you add some comments (maybe here) to explain?

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:572]: done

**Review by Ngone51 — APPROVED** [review]

**Comment by mridulm** [comment]
Thanks for fixing this @attilapiros !
Thanks for the reviews @dongjoon-hyun, @Ngone51 :-)

**Pull request closed without merging**